p2p/src/net/types/protocol.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2021-2022 RBB S.r.l
+// opensource@mintlayer.org
+// SPDX-License-Identifier: MIT
+// Licensed under the MIT License;
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://github.com/mintlayer/mintlayer-core/blob/master/LICENSE
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{
+    collections::HashSet,
+    fmt::{self, Display, Formatter},
+};
+
+use tap::TapFallible;
+
+use common::primitives::semver::SemVer;
+use logging::log;
+use serialization::{Decode, Encode};
+
+/// Protocol type and version.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Encode, Decode)]
+pub struct Protocol {
+    version: SemVer,
+    protocol: ProtocolType,
+}
+
+/// Protocol type.
+#[derive(Debug, Copy, Clone, Eq, PartialEq, Hash, Encode, Decode)]
+pub enum ProtocolType {
+    /// The publish/subscription protocol.
+    Pubsub,
+    /// The ping protocol can be used as a simple application-layer health check. Corresponds to
+    /// the "/ipfs/ping" protocol.
+    Ping,
+    /// The synchronisation protocol.
+    Sync,
+}
+
+impl Protocol {
+    /// Constructs a new protocol instance with given type and version.
+    pub const fn new(protocol: ProtocolType, version: SemVer) -> Self {
+        Self { protocol, version }
+    }
+
+    /// Parses a protocol from the given string.
+    ///
+    /// A string must contain the name of a protocol and a version after the last slash. For
+    /// example: "/ipfs/ping/1.0.0" or "/mintlayer/sync/0.1.0".
+    pub fn parse(val: &str) -> Option<Self> {
```

Comment:
I think this is in the wrong place. Parsing the backend-specific protocol string into `Protocol` should be part of the backend code. So instead of this global `parse()` there should be a `TryFrom<String> for Protocol` inside `src/net/libp2p/service/connectivity.rs` which parses the libp2p-specific protocol strings into this global and implementation-agnostic protocol type. Same thing for mock interface as well.

Reply:
Makes sens, though trait implementations are kind of global.

Do I understand correctly that mocks should have an independent implementation of protocols parsing?